File: packages/cli/hieudoanm/rust/src/libs/requests.rs
use std::collections::HashMap;
use std::time::Duration;

use crate::libs::colors;

pub const HEADER_CONTENT_TYPE: &str = "Content-Type";
pub const HEADER_ACCEPT: &str = "Accept";
pub const HEADER_AUTHORIZATION: &str = "Authorization";
pub const HEADER_USER_AGENT: &str = "User-Agent";
pub const HEADER_CACHE_CONTROL: &str = "Cache-Control";
pub const HEADER_CONTENT_LENGTH: &str = "Content-Length";
pub const HEADER_CONTENT_ENCODING: &str = "Content-Encoding";
pub const HEADER_ACCEPT_ENCODING: &str = "Accept-Encoding";
pub const HEADER_ACCEPT_LANGUAGE: &str = "Accept-Language";

pub const CONTENT_TYPE_JSON: &str = "application/json";
pub const CONTENT_TYPE_XML: &str = "application/xml";
pub const CONTENT_TYPE_FORM_URLENCODED: &str = "application/x-www-form-urlencoded";
pub const CONTENT_TYPE_TEXT_PLAIN: &str = "text/plain";
pub const CONTENT_TYPE_TEXT_HTML: &str = "text/html";
pub const CONTENT_TYPE_MULTIPART_FORM: &str = "multipart/form-data";

pub const LOG_RESPONSE_STATUS: &str = "Response Status";
pub const LOG_RESPONSE_BODY: &str = "Response Body";

#[derive(Debug, Default, Clone)]
pub struct Options {
    pub headers: HashMap<String, String>,
    pub query: HashMap<String, String>,
    pub body: Option<serde_json::Value>,
    pub timeout: Duration,
    pub retries: usize,
    pub debug: bool,
}

fn should_retry(
    err: Option<&anyhow::Error>,
    status: u16,
    attempt: usize,
    max_retries: usize,
) -> bool {
    if attempt >= max_retries {
        return false;
    }
    if err.is_some() {
        return true;
    }
    status >= 500 && status <= 599
}

fn backoff(attempt: usize) {
    std::thread::sleep(Duration::from_millis((attempt as u64 + 1) * 300));
}

async fn attempt_request(
    method: reqwest::Method,
    url: &str,
    options: &Options,
) -> Result<reqwest::Response, anyhow::Error> {
    let client = reqwest::Client::new();
    let mut req = client.request(method, url);

    for (k, v) in &options.headers {
        req = req.header(k, v);
    }
    if !options.query.is_empty() {
        req = req.query(&options.query);
    }
    if options.body.is_some() {
        req = req.json(&options.body);
        req = req.header(HEADER_CONTENT_TYPE, CONTENT_TYPE_JSON);
    }

    let resp = req.send().await?;
    Ok(resp)
}

fn log_response(resp: &reqwest::Response, body: &[u8], options: &Options) {
    if !options.debug {
        return;
    }

    const MAX_BODY_LENGTH: usize = 1000;
    let body_str = String::from_utf8_lossy(body);
    let body_str = if body_str.len() > MAX_BODY_LENGTH {
        format!("{}...[truncated]", &body_str[..MAX_BODY_LENGTH])
    } else {
        body_str.to_string()
    };

    let status = resp.status();
    let status_color = if status.is_server_error() {
        colors::red
    } else if status.is_client_error() {
        colors::yellow
    } else {
        colors::green
    };

    let headers_str = format!("{:?}", resp.headers());

    println!(
        "{}===== HTTP Response Debug ====={}",
        colors::blue(""),
        colors::blue("")
    );
    println!("Status: {}{}{}", status_color(&status.to_string()), "", "");
    println!("Headers: {headers_str}");
    println!(
        "Body: {}{}{}",
        colors::gray(""),
        colors::gray(&body_str),
        ""
    );
    println!(
        "{}==============================={}",
        colors::blue(""),
        colors::blue("")
    );
}

pub async fn do_request(
    method: reqwest::Method,
    url: &str,
    options: Options,
) -> Result<Vec<u8>, anyhow::Error> {
    let max_retries = options.retries;
    let timeout = if options.timeout.as_secs() == 0 {
        Duration::from_secs(10)
    } else {
        options.timeout
    };

    let mut last_err: Option<anyhow::Error> = None;

    for attempt in 0..=max_retries {
        let result =
            tokio::time::timeout(timeout, attempt_request(method.clone(), url, &options)).await;

        match result {
            Ok(Ok(resp)) => {
                let status = resp.status();
                let body = resp.bytes().await?.to_vec();

                if should_retry(None, status.as_u16(), attempt, max_retries) {
                    last_err = Some(anyhow::anyhow!("server error: {}", status));
                    backoff(attempt);
                    continue;
                }

                log_response_raw(&status, &body, &options);
                return Ok(body);
            }
            Ok(Err(e)) => {
                last_err = Some(anyhow::anyhow!("{e}"));
                if should_retry(Some(&last_err.as_ref().unwrap()), 0, attempt, max_retries) {
                    backoff(attempt);
                    continue;
                }
                return Err(e);
            }
            Err(_) => {
                last_err = Some(anyhow::anyhow!("timeout"));
                if should_retry(None, 0, attempt, max_retries) {
                    backoff(attempt);
                    continue;
                }
                return Err(anyhow::anyhow!("request timed out"));
            }
        }
    }

    Err(anyhow::anyhow!(
        "request failed after retries: {:?}",
        last_err
    ))
}

fn log_response_raw(status: &reqwest::StatusCode, body: &[u8], options: &Options) {
    if !options.debug {
        return;
    }
    const MAX_BODY_LENGTH: usize = 1000;
    let body_str = String::from_utf8_lossy(body);
    let body_str = if body_str.len() > MAX_BODY_LENGTH {
        format!("{}...[truncated]", &body_str[..MAX_BODY_LENGTH])
    } else {
        body_str.to_string()
    };

    let status_color = if status.is_server_error() {
        colors::red
    } else if status.is_client_error() {
        colors::yellow
    } else {
        colors::green
    };

    println!(
        "{}===== HTTP Response Debug ====={}",
        colors::blue(""),
        colors::blue("")
    );
    println!("Status: {}", status_color(&status.to_string()));
    println!(
        "Body: {}{}{}",
        colors::gray(""),
        colors::gray(&body_str),
        ""
    );
    println!(
        "{}==============================={}",
        colors::blue(""),
        colors::blue("")
    );
}

pub async fn get(url: &str, options: Options) -> Result<Vec<u8>, anyhow::Error> {
    do_request(reqwest::Method::GET, url, options).await
}

pub async fn post(url: &str, options: Options) -> Result<Vec<u8>, anyhow::Error> {
    do_request(reqwest::Method::POST, url, options).await
}

pub async fn put(url: &str, options: Options) -> Result<Vec<u8>, anyhow::Error> {
    do_request(reqwest::Method::PUT, url, options).await
}

pub async fn patch(url: &str, options: Options) -> Result<Vec<u8>, anyhow::Error> {
    do_request(reqwest::Method::PATCH, url, options).await
}

pub async fn delete(url: &str, options: Options) -> Result<Vec<u8>, anyhow::Error> {
    do_request(reqwest::Method::DELETE, url, options).await
}

File: packages/cli/hieudoanm/cobra.go/src/libs/requests/requests.go
package requests

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net"
	"net/http"
	"net/url"
	"time"
)

// ANSI color codes
const (
	ColorReset  = "\033[0m"
	ColorRed    = "\033[31m"
	ColorGreen  = "\033[32m"
	ColorYellow = "\033[33m"
	ColorBlue   = "\033[34m"
	ColorCyan   = "\033[36m"
	ColorGray   = "\033[90m"
)

// ------------------------
// Header Keys (common request headers)
// ------------------------
const (
	HEADER_CONTENT_TYPE     = "Content-Type"     // Format of the request body
	HEADER_ACCEPT           = "Accept"           // Format the client expects in the response
	HEADER_AUTHORIZATION    = "Authorization"    // Bearer/Basic/API key tokens
	HEADER_USER_AGENT       = "User-Agent"       // Identifies the client application
	HEADER_CACHE_CONTROL    = "Cache-Control"    // Controls caching behavior
	HEADER_CONTENT_LENGTH   = "Content-Length"   // Size of the request body (usually auto-set)
	HEADER_CONTENT_ENCODING = "Content-Encoding" // Encoding such as gzip/deflate
	HEADER_ACCEPT_ENCODING  = "Accept-Encoding"  // What encodings the client can accept
	HEADER_ACCEPT_LANGUAGE  = "Accept-Language"  // Preferred language for the response
)

// ------------------------
// MIME / Content-Type Values
// ------------------------
const (
	CONTENT_TYPE_JSON            = "application/json"
	CONTENT_TYPE_XML             = "application/xml"
	CONTENT_TYPE_FORM_URLENCODED = "application/x-www-form-urlencoded"
	CONTENT_TYPE_TEXT_PLAIN      = "text/plain"
	CONTENT_TYPE_TEXT_HTML       = "text/html"
	CONTENT_TYPE_MULTIPART_FORM  = "multipart/form-data"
)

// ------------------------
// Log Labels – used for debug printing
// ------------------------
const (
	LOG_RESPONSE_STATUS = "Response Status"
	LOG_RESPONSE_BODY   = "Response Body"
)

// Default HTTP client with timeout.
// Users can replace this by overriding directly if needed.
var client = &http.Client{Timeout: 15 * time.Second}

// Options configures request behavior: headers, query params,
// body payload, timeout per request, and retry behavior.
type Options struct {
	Header  http.Header
	Query   map[string]string
	Body    interface{}
	Timeout time.Duration
	Retries int
	Debug   bool
}

// ------------------------
// Small helper functions
// ------------------------

// buildURL adds query parameters to a base URL.
// Returns a parsed *url.URL or an error.
func buildURL(rawURL string, query map[string]string) (*url.URL, error) {
	u, err := url.Parse(rawURL)
	if err != nil {
		return nil, err
	}
	if query != nil {
		q := u.Query()
		for k, v := range query {
			q.Set(k, v)
		}
		u.RawQuery = q.Encode()
	}
	return u, nil
}

// buildBody marshals the request body into JSON if non-nil.
// Returns an io.Reader or nil if no body.
func buildBody(body interface{}) (io.Reader, error) {
	if body == nil {
		return nil, nil
	}
	jsonBytes, err := json.Marshal(body)
	if err != nil {
		return nil, err
	}
	return bytes.NewBuffer(jsonBytes), nil
}

// createRequest prepares the *http.Request with context, URL, body,
// and attaches all provided headers.
func createRequest(ctx context.Context, method string, u *url.URL, body io.Reader, opt Options) (*http.Request, error) {
	req, err := http.NewRequestWithContext(ctx, method, u.String(), body)
	if err != nil {
		return nil, err
	}

	// Copy custom headers
	for k, values := range opt.Header {
		for _, v := range values {
			req.Header.Add(k, v)
		}
	}

	// Automatically set JSON Content-Type if body is present
	if opt.Body != nil {
		req.Header.Set(HEADER_CONTENT_TYPE, CONTENT_TYPE_JSON)
	}

	return req, nil
}

// shouldRetry determines whether a request should be retried.
// Retries happen on:
// - network errors
// - HTTP 5xx server errors
// - only if attempts < maxRetries
func shouldRetry(err error, status int, attempt, maxRetries int) bool {
	if attempt >= maxRetries {
		return false
	}

	// Retry network errors (connection reset, timeout, etc.)
	if err != nil {
		var netErr net.Error
		return errors.As(err, &netErr)
	}

	// Retry 5xx errors
	return status >= 500 && status <= 599
}

// backoff waits before trying the next retry attempt.
func backoff(attempt int) {
	time.Sleep(time.Duration(attempt+1) * 300 * time.Millisecond)
}

// ------------------------
// Main request functions
// ------------------------

// handleResponse reads the response body safely and returns the body bytes
// and status code. Ensures resp.Body is always closed.
func handleResponse(resp *http.Response) ([]byte, int, error) {
	if resp == nil {
		return nil, 0, errors.New("nil response")
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, resp.StatusCode, err
	}

	return body, resp.StatusCode, nil
}

// attemptRequest builds the URL, encodes the body, creates the request,
// and executes it using the shared HTTP client.
func attemptRequest(
	method string,
	rawURL string,
	options Options,
	ctx context.Context,
) (*http.Response, error) {

	urlObj, err := buildURL(rawURL, options.Query)
	if err != nil {
		return nil, err
	}

	body, err := buildBody(options.Body)
	if err != nil {
		return nil, err
	}

	req, err := createRequest(ctx, method, urlObj, body, options)
	if err != nil {
		return nil, err
	}

	return client.Do(req)
}

// joinHeaderValues joins multiple header values into a single string
func joinHeaderValues(values []string) string {
	if len(values) == 0 {
		return ""
	}
	return values[0] + func() string {
		if len(values) > 1 {
			return ", " + joinHeaderValues(values[1:])
		}
		return ""
	}()
}

// logResponse prints request/response debug info with colors
func logResponse(resp *http.Response, body []byte, options Options) {
	if !options.Debug || resp == nil {
		return
	}

	// Limit body output
	const maxBodyLength = 1000
	bodyStr := string(body)
	if len(bodyStr) > maxBodyLength {
		bodyStr = bodyStr[:maxBodyLength] + "...[truncated]"
	}

	// Format headers
	var headersBuf bytes.Buffer
	for k, v := range resp.Header {
		headersBuf.WriteString(ColorCyan + k + ColorReset + ": " + joinHeaderValues(v) + "; ")
	}

	// Color status based on code
	statusColor := ColorGreen
	if resp.StatusCode >= 400 && resp.StatusCode < 500 {
		statusColor = ColorYellow
	} else if resp.StatusCode >= 500 {
		statusColor = ColorRed
	}

	log.Println(ColorBlue + "===== HTTP Response Debug =====" + ColorReset)
	log.Printf("Status: %s%s%s\n", statusColor, resp.Status, ColorReset)
	log.Printf("Headers: %s\n", headersBuf.String())
	log.Printf("Body: %s%s%s\n", ColorGray, bodyStr, ColorReset)
	log.Println(ColorBlue + "===============================" + ColorReset)
}

// doRequest performs the full request cycle including:
// - timeout control
// - retry logic
// - response handling
// - logging (status + body)
func doRequest(method, rawURL string, options Options) ([]byte, error) {
	maxRetries := options.Retries

	// Default timeout if not provided
	timeout := options.Timeout
	if timeout == 0 {
		timeout = 10 * time.Second
	}

	var lastErr error

	for attempt := 0; attempt <= maxRetries; attempt++ {

		// Each attempt gets its own context with timeout
		ctx, cancel := context.WithTimeout(context.Background(), timeout)

		resp, err := attemptRequest(method, rawURL, options, ctx)
		cancel()

		// Network error or connection failure
		if err != nil {
			lastErr = err

			if shouldRetry(err, 0, attempt, maxRetries) {
				backoff(attempt)
				continue
			}
			return nil, err
		}

		// Read and process the response
		body, status, readErr := handleResponse(resp)
		if readErr != nil {
			return nil, readErr
		}

		// Retry server failures (HTTP 5xx)
		if shouldRetry(nil, status, attempt, maxRetries) {
			lastErr = fmt.Errorf("server error: %v", resp.Status)
			backoff(attempt)
			continue
		}

		// Debug logging
		logResponse(resp, body, options)

		return body, nil
	}

	return nil, fmt.Errorf("request failed after retries: %w", lastErr)
}

// Public convenience wrappers for common HTTP verbs
func Get(url string, options Options) ([]byte, error) { return doRequest(http.MethodGet, url, options) }
func Post(url string, options Options) ([]byte, error) {
	return doRequest(http.MethodPost, url, options)
}
func Put(url string, options Options) ([]byte, error) { return doRequest(http.MethodPut, url, options) }
func Patch(url string, options Options) ([]byte, error) {
	return doRequest(http.MethodPatch, url, options)
}
func Delete(url string, options Options) ([]byte, error) {
	return doRequest(http.MethodDelete, url, options)
}
